src/uct/ib/base/ib_iface.c

```diff
@@ -406,60 +407,79 @@ ucs_status_t uct_ib_iface_get_device_address(uct_iface_h tl_iface,
+    remote_roce_ver = uct_ib_address_flags_get_roce_version(remote_ib_addr_flags);
+    ucs_assert(remote_roce_ver < UCT_IB_DEVICE_ROCE_VER_ERR);
 
-    if (!ret) {
-        ucs_debug("different RoCE versions detected. local %s remote %s",
-                  uct_ib_roce_version_str(local_roce_ver->ver),
-                  uct_ib_roce_version_str(remote_roce_ver));
+    if (local_roce_info->ver == remote_roce_ver) {
+        return 1;
     }
 
-    return ret;
+    ucs_trace("different RoCE versions detected. local %s (gid=%s)"
+              "remote %s (gid=%s)",
+              uct_ib_roce_version_str(local_roce_info->ver),
+              uct_ib_device_gid_ipv6_str(&local_gid_info->gid, gid_str,
+                                         sizeof(gid_str)),
+              uct_ib_roce_version_str(remote_roce_ver),
+              uct_ib_device_gid_ipv6_str((union ibv_gid *)remote_ib_addr_gid,
+                                         gid_str, sizeof(gid_str)));
+    return 0;
 }
 
-static int uct_ib_iface_is_same_roce_info(uct_ib_roce_version_desc_t *local_roce_ver,
+static int uct_ib_iface_is_same_roce_info(uct_ib_device_gid_info_t *local_gid_info,
                                           const uct_ib_address_t *remote_ib_addr)
 {
-    if ((remote_ib_addr->flags >> ucs_ilog2(UCT_IB_ADDRESS_FLAG_LAST)) ==
+    uct_ib_roce_version_info_t *local_roce_info = &local_gid_info->roce_info;
+    uint8_t remote_ib_addr_flags                = remote_ib_addr->flags;
+    sa_family_t remote_ib_addr_af;
+
+    if ((uct_ib_address_flags_get_roce_version(remote_ib_addr_flags)) ==
          UCT_IB_DEVICE_ROCE_ANY) {
         return 1;
     }
 
-    if ((local_roce_ver->addr_family == AF_INET6) &&
-        (remote_ib_addr->flags & UCT_IB_ADDRESS_FLAG_ROCE_IPV6)) {
-        return uct_ib_iface_is_same_roce_version(local_roce_ver, remote_ib_addr);
-    }
+    remote_ib_addr_af = (remote_ib_addr_flags & UCT_IB_ADDRESS_FLAG_ROCE_IPV6) ?
+                         AF_INET6 : AF_INET;
 
-    if ((local_roce_ver->addr_family == AF_INET) &&
-        !(remote_ib_addr->flags & UCT_IB_ADDRESS_FLAG_ROCE_IPV6)) {
-        return uct_ib_iface_is_same_roce_version(local_roce_ver, remote_ib_addr);
+    if (local_roce_info->addr_family != remote_ib_addr_af) {
+        ucs_assert(local_roce_info->addr_family != 0);
+        ucs_debug("different addr_family detected. local %s remote %s",
+                  ucs_sockaddr_address_family_str(local_roce_info->addr_family),
+                  ucs_sockaddr_address_family_str(remote_ib_addr_af));
+       return 0;
     }
 
-    ucs_assert(local_roce_ver->addr_family != 0);
-    ucs_debug("different addr_family detected. local %s remote %s",
-              (local_roce_ver->addr_family == AF_INET6 ? "AF_INET6" : "AF_INET"),
-              (remote_ib_addr->flags & UCT_IB_ADDRESS_FLAG_ROCE_IPV6 ?
-               "AF_INET6" : "AF_INET4"));
-    return 0;
+    return uct_ib_iface_roce_is_reachable(local_gid_info, remote_ib_addr);
```

Comment:
i think we can unite uct_ib_iface_is_same_roce_info and uct_ib_iface_roce_is_reachable to one function - uct_ib_iface_roce_is_reachable()

Reply:
it used to be one function but i split it during CR. 
you mean to have a function that checks both addr_family and roce version one after the other?

Reply:
yes